crates/ide-completion/src/render/function.rs

```diff
@@ -278,31 +278,44 @@ pub(super) fn add_call_parens<'b>(
-        let in_match_arm = ctx.token.parent_ancestors().try_for_each(|ancestor| {
-            if ast::MatchArm::can_cast(ancestor.kind()) {
-                ControlFlow::Break(true)
-            } else if matches!(ancestor.kind(), SyntaxKind::EXPR_STMT | SyntaxKind::BLOCK_EXPR) {
-                ControlFlow::Break(false)
-            } else {
-                ControlFlow::Continue(())
+        let inside_closure_ret = ctx.token.parent_ancestors().try_for_each(|ancestor| {
+            match_ast! {
+                match ancestor {
+                    ast::BlockExpr(_) => ControlFlow::Break(false),
+                    ast::ClosureExpr(_) => ControlFlow::Break(true),
+                    _ => ControlFlow::Continue(())
+                }
             }
         });
-        // FIXME: This will assume expr macros are not inside match, we need to somehow go to the "parent" of the root node.
-        let in_match_arm = match in_match_arm {
-            ControlFlow::Continue(()) => false,
-            ControlFlow::Break(it) => it,
-        };
-        let complete_token = if in_match_arm { T![,] } else { T![;] };
-        if next_non_trivia_token.map(|it| it.kind()) != Some(complete_token) {
-            cov_mark::hit!(complete_semicolon);
-            let ch = if in_match_arm { ',' } else { ';' };
-            if snippet.ends_with("$0") {
-                snippet.insert(snippet.len() - "$0".len(), ch);
-            } else {
-                snippet.push(ch);
+
+        if inside_closure_ret != ControlFlow::Break(true) {
+            let next_non_trivia_token =
+                std::iter::successors(ctx.token.next_token(), |it| it.next_token())
+                    .find(|it| !it.kind().is_trivia());
+            let in_match_arm = ctx.token.parent_ancestors().try_for_each(|ancestor| {
+                if ast::MatchArm::can_cast(ancestor.kind()) {
+                    ControlFlow::Break(true)
+                } else if matches!(ancestor.kind(), SyntaxKind::EXPR_STMT | SyntaxKind::BLOCK_EXPR)
+                {
+                    ControlFlow::Break(false)
+                } else {
+                    ControlFlow::Continue(())
+                }
+            });
+            // FIXME: This will assume expr macros are not inside match, we need to somehow go to the "parent" of the root node.
+            let in_match_arm = match in_match_arm {
+                ControlFlow::Continue(()) => false,
+                ControlFlow::Break(it) => it,
+            };
+            let complete_token = if in_match_arm { T![,] } else { T![;] };
+            if next_non_trivia_token.map(|it| it.kind()) != Some(complete_token) {
+                cov_mark::hit!(complete_semicolon);
+                let ch = if in_match_arm { ',' } else { ';' };
+                if snippet.ends_with("$0") {
+                    snippet.insert(snippet.len() - "$0".len(), ch);
+                } else {
+                    snippet.push(ch);
+                }
```

Comment:
Likewise for tihs (apart form what depends on the snippet / function to be rendered)